File: src/cdisc_library_api/specs/ct_package.clj
(ns cdisc-library-api.specs.ct-package
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [cdisc-library-api.specs.ct-package-links :refer :all]
            [cdisc-library-api.specs.ct-package-codelists :refer :all]
            )
  (:import (java.io File)))


(def ct-package-data
  {
   (ds/opt :name) string?
   (ds/opt :label) string?
   (ds/opt :description) string?
   (ds/opt :source) string?
   (ds/opt :effectiveDate) string?
   (ds/opt :registrationStatus) string?
   (ds/opt :version) string?
   (ds/opt :_links) ct-package-links-spec
   (ds/opt :codelists) (s/coll-of ct-package-codelists-spec)
   })

(def ct-package-spec
  (ds/spec
    {:name ::ct-package
     :spec ct-package-data}))

File: src/main/scala/org/openapitools/client/api/CtPackage.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import CtPackage._

case class CtPackage (
  name: Option[String],
label: Option[String],
description: Option[String],
source: Option[String],
effectiveDate: Option[String],
registrationStatus: Option[String],
version: Option[String],
links: Option[CtPackageLinks],
codelists: Option[List[CtPackageCodelists]])

object CtPackage {
  import DateTimeCodecs._

  implicit val CtPackageCodecJson: CodecJson[CtPackage] = CodecJson.derive[CtPackage]
  implicit val CtPackageDecoder: EntityDecoder[CtPackage] = jsonOf[CtPackage]
  implicit val CtPackageEncoder: EntityEncoder[CtPackage] = jsonEncoderOf[CtPackage]
}
